meshio/gmsh/_gmsh41.py

```diff
@@ -307,52 +371,210 @@ def write4_1(filename, mesh, float_fmt=".16e", binary=True):
+    #      numSurfaces(size_t) numVolumes(size_t)
+    #    pointTag(int) X(double) Y(double) Z(double)
+    #      numPhysicalTags(size_t) physicalTag(int) ...
+    #    ...
+    #    curveTag(int) minX(double) minY(double) minZ(double)
+    #      maxX(double) maxY(double) maxZ(double)
+    #      numPhysicalTags(size_t) physicalTag(int) ...
+    #      numBoundingPoints(size_t) pointTag(int) ...
+    #    ...
+    #    surfaceTag(int) minX(double) minY(double) minZ(double)
+    #      maxX(double) maxY(double) maxZ(double)
+    #      numPhysicalTags(size_t) physicalTag(int) ...
+    #      numBoundingCurves(size_t) curveTag(int) ...
+    #    ...
+    #    volumeTag(int) minX(double) minY(double) minZ(double)
+    #      maxX(double) maxY(double) maxZ(double)
+    #      numPhysicalTags(size_t) physicalTag(int) ...
+    #      numBoundngSurfaces(size_t) surfaceTag(int) ...
+
+    # Both nodes and cells have enities, but the cell entities are a subset of
+    # the nodes. The reason is (if the inner workings of Gmsh has been correctly
+    # understood) that node entities are assigned to all
+    # objects necessary to specify the geometry whereas only cells of Physical
+    # objcets (gmsh jargon) are present among the cell entities.
+    # The entities section must therefore be built on the node-entities, if
+    # these are available. If this is not the case, we leave this section blank.
+    # TODO: Should this give a warning?
+    if "gmsh:dim_tags" not in point_data:
+        return
+
+    fh.write(b"$Entities\n")
+
+    # Array of entity tag (first row) and dimension (second row) per node.
+    # We need to combine the two, since entity tags are reset for each dimension.
+    # Uniquify, so that each row in node_dim_tags represent a unique entity
+    node_dim_tags = numpy.unique(point_data["gmsh:dim_tags"], axis=0)
+
+    # Write number of entities per dimension
+    num_occ = numpy.bincount(node_dim_tags[:, 0], minlength=4)
+    if binary:
+        num_occ.astype(c_size_t).tofile(fh)
+    else:
+        fh.write(
+            f"{num_occ[0]} {num_occ[1]} {num_occ[2]} {num_occ[3]}\n".encode("utf-8")
+        )
+
+    # Array of dimension and entity tag per cell. Will be compared with the
+    # similar not array.
+    cell_dim_tags = numpy.empty((len(cells), 2), dtype=int)
+    for ci in range(len(cells)):
+        cell_dim_tags[ci] = [
+            _geometric_dimension[cells[ci][0]],
+            cell_data["gmsh:geometrical"][ci][0],
+        ]
+
+    # We will only deal with bounding entities if this information is available
+    has_bounding_elements = "gmsh:bounding_entities" in cell_sets
+
+    # The node entities form a superset of cell entities. Write entity information
+    # based on nodes, supplement with cell information when there is a matcihng
+    # cell block.
+    for dim, tag in node_dim_tags:
+
+        # Find the matching cell block, if it exists
+        matching_cell_block = numpy.where(
```

Comment:
We can probably live without `where` here. Just store the boolean array and replace `matching_cell_block.size > 0` by `numpy.any(matching_cell_block)`.

Reply:
If matching_cell_block is boolean, we still need the index for the match (see calls `cell_data["gmsh:physical"][matching_cell_blocks[0]]` below).

Reply:
Okay, yes.